javalib/src/main/scala/java/util/TreeSet.scala

```diff
@@ -0,0 +1,222 @@
+package java.util
+
+import java.lang.Comparable
+import scala.collection.mutable
+import scala.math.Ordering
+
+import scala.collection.JavaConversions._
+
+class TreeSet[E] protected (ordering: Ordering[_ >: E], _comparator: Comparator[_ >: E])
+    extends AbstractSet[E]
+    with NavigableSet[E]
+    with Cloneable
+    with Serializable { self =>
+
+  def this() =
+    this(defaultOrdering[E], null.asInstanceOf[Comparator[_ >: E]])
+
+  def this(comparator: Comparator[_ >: E]) =
+    this(Ordering.comparatorToOrdering(comparator), null.asInstanceOf[Comparator[E]])
+
+  def this(collection: Collection[_ <: E]) = {
+    this(defaultOrdering[E], null.asInstanceOf[Comparator[E]])
+    addAll(collection)
+  }
+
+  def this(sortedSet: SortedSet[E]) = {
+    this(Ordering.comparatorToOrdering(sortedSet.comparator()),
+        sortedSet.comparator())
+    addAll(sortedSet)
+  }
+
+  private implicit object BoxOrdering extends Ordering[Box[E]] {
+    def compare(a: Box[E], b:Box[E]): Int = ordering.compare(a.inner, b.inner)
+  }
+
+  private val inner: mutable.TreeSet[Box[E]] = new mutable.TreeSet[Box[E]]()
+
+  def iterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.clone.iterator
```

Comment:
You should not clone a `mutable.TreeSet`. You could reset the iterator after a `remove` with `inner.iteratorFrom`.